tests/binary_agreement_mitm.rs

```diff
@@ -0,0 +1,210 @@
+#![deny(unused_must_use)]
+//! Tests the BinaryAgreement protocol with a MTIM adversary.
+
+extern crate env_logger;
+extern crate failure;
+extern crate hbbft;
+extern crate integer_sqrt;
+extern crate proptest;
+extern crate rand;
+extern crate threshold_crypto;
+
+pub mod net;
+
+use std::iter;
+use std::sync::Arc;
+
+use hbbft::binary_agreement::{BinaryAgreement, Message, MessageContent, SbvMessage};
+use hbbft::{DistAlgorithm, Step};
+
+use net::adversary::{NetMutHandle, QueuePosition};
+use net::err::CrankError;
+use net::{Adversary, NetBuilder, NetMessage};
+
+type NodeId = usize;
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+enum ReorderingStage {
+    DeliverBBVal,
+    DeliverA0BVal,
+    DeliverA0Aux,
+    Finished,
+}
+
+impl Default for ReorderingStage {
+    fn default() -> Self {
+        ReorderingStage::DeliverBBVal
+    }
+}
+
+/// An adversary for the reordering attack.
+/// Described here: https://github.com/amiller/HoneyBadgerBFT/issues/59#issue-310368284
+/// Excluding the first node, which is F,
+/// A0 is the first third of nodes, A1 is the second third, and the rest are B.
+#[derive(Default)]
+struct ReorderingAdversary {
+    stage: ReorderingStage,
+    stage_progress: usize,
+    sent_initial_messages: bool,
+    sent_final_messages: bool,
+}
+
+const NODES_PER_GROUP: usize = 10;
```

Comment:
Can this value be randomized? For example, sampled randomly from a given range?

Reply:
On the other hand this is just a regression test for one specific case. I wonder whether we should just set this to `1`?

Reply:
In that case, making it random will answer both questions :P